### PR TITLE
Add `strip_delimiters` option to `read_text`

### LINKED ISSUE #11625: [FEA] `read_text` should support removing delimiters from the output
**Is your feature request related to a problem? Please describe.**
At the moment, the string column generated by `read_text` uses the underlying input bytes verbatim without removing the delimiter entries.

**Describe the solution you'd like**
This could be addressed by adding a `remove_delimiters` bool parameter to the function. This would only require one additional transform + scan to compute the output offsets without delimiters and one string gather to copy the bytes. Since we are far from peak memory bandwidth, it shouldn't matter much performance-wise.

**Describe alternatives you've considered**
The copy operation could also happen directly inside the `multibyte_split` kernel, which would avoid the need for a `gather` step, but add a third scan operation to the kernel.

### PR DESCRIPTION
**Pull request by upsj**
## Description

This adds a `strip_delimiters` post-processing option to `read_text`. I needed to implement some lightweight striping because a thread-per-row parallelization of the string gather gave pretty bad performance.

For consistency, I also removed the special-case handling of delimiters at the end (previously adding an empty row), to match the read_csv behavior.

Benchmark results:

```
benchmarks/MULTIBYTE_SPLIT_NVBENCH --axis size_approx[pow2]=30 --axis byte_range_percent=100 --axis T=device --axis delim_size=4
```

### [0] Tesla T4

|   T    | strip_delimiters | delim_percent |    size_approx    |   CPU Time  | Noise | Peak Memory Usage | Encoded file size |
|--------|------------------|---------------|-------------------|------------|-------|-------------------|-------------------|
| device |                0 |             1 | 2^30 = 1073741824 | 178.133 ms | 0.36% |         3.709 GiB |      1014.442 MiB |
| device |                1 |             1 | 2^30 = 1073741824 | 188.328 ms | 0.31% |         4.690 GiB |      1014.442 MiB |
| device |                0 |            25 | 2^30 = 1073741824 | 206.188 ms | 0.03% |         5.292 GiB |       953.075 MiB |
| device |                1 |            25 | 2^30 = 1073741824 | 242.534 ms | 0.50% |         5.975 GiB |       953.075 MiB |

Closes #11625

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by davidwendt** [comment]
> ... thread-per-row parallelization of the string gather gave pretty bad performance

Can you elaborate on this? I think there is a fast strings gather that may be possible to use here.

**Comment by upsj** [comment]
> Can you elaborate on this? I think there is a fast strings gather that may be possible to use here.

When running `for_each_n` with one thread per row copying the elements excluding the delimiter, I got runtimes around 1s instead of 20 ms with the current solution. NSight Compute profiles showed horrible cache utilization numbers. `cudf::strings::detail::gather` looks _almost_ right - I need to transform individual strings to strip off a fixed number of characters (except for the last row), not gather a subset of strings from a larger column. Maybe the word _gather_ is a bit misleading. But the underlying kernels look pretty close to what I need.

**Comment by davidwendt** [comment]
> > Can you elaborate on this? I think there is a fast strings gather that may be possible to use here.
> 
> When running `for_each_n` with one thread per row copying the elements excluding the delimiter, I got runtimes around 1s instead of 20 ms with the current solution. NSight Compute profiles showed horrible cache utilization numbers. `cudf::strings::detail::gather` looks _almost_ right - I need to transform individual strings to strip off a fixed number of characters (except for the last row), not gather a subset of strings from a larger column. Maybe the word _gather_ is a bit misleading. But the underlying kernels look pretty close to what I need.

Ok. I was thinking more along the lines of this `make_strings_column` function that can take a device-span of `string_view` (or equivalent `thrust::pair`): https://docs.rapids.ai/api/libcudf/stable/group__column__factories.html#ga993941cbf14270bcea2cc95427996de1

It would just be a matter of building a device-uvector of these to call one of these factory functions which has a highly tuned gather operation for building a strings column from individual strings in device memory. 

For reference, both of these factory functions (`string_view` and `thrust_pair`) call into https://github.com/rapidsai/cudf/blob/6ca2ceb8e200d55f1f681a4ca086614a28d67ad1/cpp/include/cudf/strings/detail/strings_column_factories.cuh#L72-L76

**Comment by upsj** [comment]
@davidwendt thanks for the details, that shaved another 18ms off the runtime for the long string case (at the cost of maybe 20 ms for the short string case, but I'll take the added simplicity :) )

**Comment by upsj** [comment]
rerun tests

File: python/cudf/cudf/io/text.py [hunk 1]
```diff
@@ -14,6 +14,7 @@ def read_text(
     filepath_or_buffer,
     delimiter=None,
     byte_range=None,
+    strip_delimiters=False,
```
Comment (galipremsagar) [inline python/cudf/cudf/io/text.py:17]: Can we also update the docstring with `strip_delimiters` at https://github.com/rapidsai/cudf/blob/branch-22.12/python/cudf/cudf/utils/ioutils.py#L1205?

Comment (upsj) [reply python/cudf/cudf/io/text.py:17]: done

File: cpp/include/cudf/io/text/multibyte_split.hpp [hunk 2]
```diff
@@ -30,11 +30,25 @@ namespace cudf {
...
+/**
+ * @brief Parsing options for multibyte_split.
+ */
+struct parse_options {
+  /**
+   * @brief Only rows starting inside this byte range will be part of the output column.
+   */
+  byte_range_info byte_range = create_byte_range_info_max();
+  /**
+   * @brief Should delimiters at the end of rows be stripped from the output column?
```
Comment (davidwendt) [inline cpp/include/cudf/io/text/multibyte_split.hpp:42]:
```suggestion
   * @brief Should delimiters at the end of rows be stripped from the output column
```
The question mark seems wrong to me here. Perhaps this could be rephrased otherwise I think removing the ? is fine.

**Review by davidwendt — APPROVED** [review]

File: python/cudf/cudf/io/text.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/text.py:17]: I don't see the commit that changes `ioutils.py`(link above), it could be a github issue..could you send me the link to your commit if you have already made the change?

**Review by galipremsagar — APPROVED** [review]

File: python/cudf/cudf/io/text.py (hunk 1 above)
Comment (upsj) [reply python/cudf/cudf/io/text.py:17]: Only forgot to push after the commit :)

File: cpp/include/cudf/io/text/multibyte_split.hpp (hunk 2 above)
Comment (bdice) [reply cpp/include/cudf/io/text/multibyte_split.hpp:42]: I'd rephrase.
```suggestion
   * @brief Whether delimiters at the end of rows should be stripped from the output column
```

File: python/cudf/cudf/utils/ioutils.py [hunk 3]
```diff
@@ -1215,6 +1215,8 @@
 delimiter : string, default None
     The delimiter that should be used for splitting text chunks into
     separate cudf column rows. The delimiter may be one or more characters.
+strip_delimiters : boolean, default False
+    If True, removes delimiters that are normally kept at the end of strings.
```
Comment (bdice) [inline python/cudf/cudf/utils/ioutils.py:1219]: I don't think this docstring is detailed enough to explain the behavior. It's hard to put a real example in the parameters section, but maybe we can provide more info below?

Should this be named `strip_trailing_delimiters` to clarify that it acts at the end of the input line? (or entire input file? unclear) Python's default is to strip both sides:
```python
assert "asdfa".strip("a") == "sdf"
```

**Review by bdice — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/utils/ioutils.py (hunk 3 above)
Comment (bdice) [reply python/cudf/cudf/utils/ioutils.py:1219]: Short docstrings are usually okay when we are matching a pandas API, because users have some expectations for how the pandas API might work. Since `read_text` is not aligned with a pandas API, detailed descriptions and examples are important for setting user expectations.

Comment (upsj) [reply python/cudf/cudf/utils/ioutils.py:1219]: Agreed, we also don't mention anywhere that the delimiters are kept at the end. I'll extend it a bit

Comment (upsj) [reply python/cudf/cudf/utils/ioutils.py:1219]: WDYT?
```suggestion
    Unlike the `str.split()` function, `read_text` preserves the delimiter
    at the end of a field in output by default, meaning `a;b;c` will turn into
    `['a;','b;','c']` when using `;` as a delimiter.
    Setting this option to `True` will strip these trailing delimiters,
    leaving only the contents between delimiters in the resulting column:
    `['a','b','c']`
```

Comment (bdice) [reply python/cudf/cudf/utils/ioutils.py:1219]: That is very helpful info! I’m happy with that.

**Comment by upsj** [comment]
rerun tests

**Comment by upsj** [comment]
@gpucibot merge
